### ISSUE TITLE
Performance issue with active encryption

### ISSUE THREAD

**Issue by javierfreire** [hivedb/hive#137]
**Steps to Reproduce**
When I try save an image on a encrypted box, the performance is very low.

I found 2 issues:

- Encryption is slow with big arrays:
    - Length 100000 => 100 ms
    - Length 800000 => 13000 ms

- BinaryWriterImpl multiply per 8 the size of one array
    

**Code sample**
```
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:hive/src/binary/binary_writer_impl.dart';
import 'package:hive/src/crypto_helper.dart';

void main() async {

    test("bug", () async {
        final customKey = [122, 236, 96, 98, 34, 34, 210, 137, 87, 46, 162, 22, 53, 76, 206, 206, 170, 19, 124, 133, 206, 57, 209, 241, 253, 30, 234, 166, 240, 164, 75, 62];
        final customData = List<int>();

        for (int i = 0; i < 100000; i++) {
            customData.add(i % 255);
        }

        final crypto = CryptoHelper(Uint8List.fromList(customKey));

        eval("Encrypted directly", () => crypto.encrypt(Uint8List.fromList(customData)));

        final writedBytes = eval("Write in writer", () {
            final writer = BinaryWriterImpl(null)..write(customData);
            return writer.toBytes();
        });

        eval("Encrypted writed bytes", () => crypto.encrypt(writedBytes));
    });

}

Uint8List eval(String message, Uint8List Function() process) {
    var initTime = DateTime.now();

    final result = process();
    final totalTime = DateTime.now().millisecondsSinceEpoch - initTime.millisecondsSinceEpoch;

    print("$message :: length ${result.length} in $totalTime ms");

    return result;
}
```

**Version**
 - Platform: Linux
 - Flutter version: 1.10.7
 - Hive version: 1.1.1

**Comment by leisim** [comment]
The problem is that `customData` is a `List<int>` and not `Uint8List`. Hive will encode each byte as 64-bit int which is super slow. If you use `Uint8List`, it should work.

**Comment by javierfreire** [comment]
Thank you
I'm trying encrypt images and [steel_crypt](https://pub.dev/packages/steel_crypt) has a good performance.

**Comment by leisim** [comment]
It probably uses native APIs for encryption.

**Comment by leisim** [comment]
I don't think there is anything I can do without depending on native code. Hopefully, Dart NNBD will improve the performance a little.

**Comment by AKushWarrior** [comment]
@leisim I'm the author of steel_crypt. Steel_crypt has no native APIs, and is 100% dart. It's also based off a forked implementation of pointycastle, which Hive used to depend on.

If you want, I can take a look at the encryption code...

**Comment by leisim** [comment]
@AKushWarrior 
As far as I can see, steel_crypt is just a wrapper around PointyCastle. The old Hive version from this issue used PointyCastle. With the latest Hive release, I implemented the encryption myself and in my tests, it is much faster than PointyCastle.

**Comment by AKushWarrior** [comment]
@leisim It is a wrapper around PointyCastle's AES, which I haven't touched. That's why I found it surprising that @javierfreire found steel_crypt to be fast. PointyCastle is pretty inefficient at most things, which I'm in the process of fixing. 

I ran my own benchmarks, and also found what you did. At short to intermediate length Strings, your solution works ~2-3 times faster than PointyCastle. When length of the Uint8List crosses about 50000, HiveAesCipher slows down significantly relative to PointyCastle.

I have no idea why this is, but it's probably what @javierfreire was referring to.
